src/solver/conv_cellfft.cpp

```diff
@@ -0,0 +1,1166 @@
+    p.grid_y      = (pny + p.tile_y - p.cny) / p.tile_y;
+    p.grid_x      = p.grid_x == 0 ? 1 : p.grid_x;
+    p.grid_y      = p.grid_y == 0 ? 1 : p.grid_y;
+    p.m           = pnc;
+    p.n           = qnc;
+    p.k           = bs * p.grid_x * p.grid_y;
+    uint32_t ek   = PSIZE(p.k, 7);
+    p.lda         = PSIZE(p.m, 31) >> 5;
+    p.ldb         = PSIZE(p.n, 31) >> 5;
+    p.lda         = (p.lda + (1 ^ (p.lda & 1))) << 5;
+    p.ldb         = (p.ldb + (1 ^ (p.ldb & 1))) << 5;
+    p.abks        = p.lda * ek + 16;
+    p.bbks        = p.ldb * ek + 16;
+    p.cbks        = p.lda * p.n + 16;
+    p.aldx        = p.anx * p.any;
+    p.bldx        = p.bnx * p.bny;
+    p.cldx        = p.cnx * p.cny;
+    p.aldy        = p.m * p.aldx;
+    p.bldy        = p.n * p.bldx;
+    p.cldy        = p.m * p.cldx;
+    if((p.grid_x | p.grid_y) != 1)
+    {
+        uint32_t pk   = PSIZE(p.k, 15);
+        uint32_t reso = p.grid_x * p.grid_y;
+        p.xmag        = idiv_magic(pk, reso);
+        p.ymag        = idiv_magic(reso, p.grid_x);
+    }
+}
+static void lk_cgemm(const Handle& h,
+                     const Kernel& kern,
+                     const cellfft_param_t& p,
+                     void* c,
+                     void* a,
+                     void* b,
+                     float alpha)
+{
+    // coef=alpha*(1.0/(cell_size*cell_size)), cell_size=16 or 32
+    float coef = alpha * (p.id == 0 ? 0.00390625f : 0.0009765625f);
+    h.Run(kern)(c, p.lda, p.cbks, a, b, p.lda, p.ldb, p.m, p.n, p.k, p.abks, p.bbks, coef);
+}
+static void lk_fft2d_r2c_perm_a(
+    const Handle& h, const Kernel& kern, const cellfft_param_t& p, void* dst, const void* src)
+{
+    h.Run(kern)(
+        dst, p.lda, p.abks, src, (p.dir != 2 ? 0 : 0x80000000) | p.m, p.anx, p.aldx, p.aldy);
+}
+static void lk_fft2d_r2c_perm_b(
+    const Handle& h, const Kernel& kern, const cellfft_param_t& p, void* dst, const void* src)
+{
+    h.Run(kern)(
+        dst, p.ldb, p.bbks, src, (p.dir == 0 ? 0 : 0x80000000) | p.n, p.bnx, p.bldx, p.bldy);
+}
+static void lk_fft2d_r2c_perm_pad(
+    const Handle& h, const Kernel& kern, const cellfft_param_t& p, void* dst, const void* src)
+{
+    h.Run(kern)(dst,
+                p.lda,
+                p.abks,
+                src,
+                p.m,
+                (p.dir != 2 ? 0 : 0x80000000) | p.pad_l,
+                p.aldx,
+                p.aldy,
+                p.anx,
+                p.any);
+}
+static void lk_fft2d_r2c_perm_s(
+    const Handle& h, const Kernel& kern, const cellfft_param_t& p, void* dst, const void* src)
+{
+    uint32_t ldr = p.bnx * p.bny * (p.dir == 0 ? p.k : p.n);
+    h.Run(kern)(dst, p.ldb, p.bbks, src, p.n, ldr);
+}
+static void lk_fft2d_r2c_grid_perm(
+    const Handle& h, const Kernel& kern, const cellfft_param_t& p, void* dst, const void* src)
+{
+    uint32_t grid = (p.grid_y << 16) | p.grid_x;
+    uint32_t tile = (p.tile_y << 16) | p.tile_x;
```

Comment:
Are we quite sure that grid and tile wouldn't exceed 65535? The same wrt other `<< 16` computations.

Reply:
yes, we are pretty sure. it's also a way to save SGPR usage.